### PR TITLE
Use `available_channel_type` table data to create channelForm

### PR DESCRIPTION
**Pull request by sukhwinder33445**
## Requires:

- https://github.com/Icinga/icinga-notifications/pull/108
- https://github.com/Icinga/icinga-notifications/pull/142

### CODE REVIEW

**Review by nilmerg — CHANGES_REQUESTED** [review]
Please rebase and fix static analysis errors.

File: application/forms/ChannelForm.php [hunk 1]
```diff
@@ -41,114 +57,27 @@ protected function assemble()
...
         $this->addElement(
             'select',
             'type',
             [
                 'label'    => $this->translate('Type'),
                 'class'    => 'autosubmit',
                 'required' => true,
-                'disable'  => [''],
-                'options'  => $type
+                'disabledOptions'  => [''],
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:67]: Either indent all `=>` to the same level, or all with a single space before and after.

File: application/forms/ChannelForm.php [hunk 2]
```diff
@@ -21,10 +30,17 @@ class ChannelForm extends CompatForm
...
     {
         $this->db = $db;
         $this->channelId = $channelId;
+
+        /** @var GettextTranslator $translateInstance */
+        $translateInstance = StaticTranslator::$instance;
+        $this->locale = $translateInstance->getLocale();
     }
 
     protected function assemble()
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:46]: please fetch from available_channel_type only once here. And default to the first found type already. There's no need for the user to specifically choose one.

File: application/forms/ChannelForm.php [hunk 3]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
             return;
         }
 
-        $channel = [
-            'name' => $this->getValue('name'),
-            'type' => $this->getValue('type')
+        $channel = $this->getValues();
+        $channel['config'] = json_encode($this->getValue('config'));
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:160]: `$channel['config']`? Instead of `getValue('config')`?

File: application/forms/ChannelForm.php [hunk 4]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
...
+        /** @var string $configs */
+        $configs = AvailableChannelType::on($this->db)
+            ->filter(Filter::equal('type', $type))
+            ->columns('config_attrs')
+            ->first()
+            ->config_attrs;
+
+        /** @var array<int, stdClass> $elementsConfig */
+        $elementsConfig = json_decode($configs, false);
+        var_dump($elementsConfig);
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:184]: why is it so hard to not commit something like this?

File: application/forms/ChannelForm.php [hunk 5]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
...
+        if (empty($elementsConfig)) {
+            return;
+        }
+
+        /** @var FormElementDecorator $decorator */
+        $decorator = $this->getDefaultElementDecorator();
+        $config = (new FieldsetElement('config'))
+            ->setDefaultElementDecorator($decorator);
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:192]: I thought fieldsets propagate that automatically? Why is this required?

File: application/forms/ChannelForm.php [hunk 6]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
...
-                'host' => $this->getValue('host'),
-                'port' => $this->getValue('port'),
-                'from' => $this->getValue('from'),
-                'password' => $this->getValue('password')
-            ];
-            if ($this->getElement('tls')->isChecked()) {
-                $channel['config']['tls'] = true;
-                $channel['config']['tls_certcheck'] = $this->getValue('tls_certcheck');
+        if (isset($elementConfig->help)) {
+            $options['title'] = $this->fromCurrentLocale($elementConfig->help);
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:263]: That should be `description`, not `title`.

File: application/forms/ChannelForm.php [hunk 7]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
...
+        }
+
+        if (isset($elementConfig->required)) {
+            $options['required'] = $elementConfig->required;
+        }
+
+        if (isset($elementConfig->placeholder)) {
+            $options['placeholder'] = $elementConfig->placeholder;
+        }
+
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:273]: remove that, that's what `default` is for

File: application/forms/ChannelForm.php [hunk 8]
```diff
@@ -229,35 +156,162 @@ protected function onSuccess()
...
+        if (isset($elementConfig->default)) {
+            $options['value'] = $elementConfig->default;
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:275]: should be `placeholder`

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Notifications/Model/Channel.php [hunk 9]
```diff
@@ -55,31 +56,43 @@ public function createRelations(Relations $relations)
...
     /**
-     * Fetch and map all the configured channel types to a key => value array
+     * Get the channel icon
+     *
+     * @return Icon
+     */
+    public function getIcon(): Icon
+    {
+        switch ($this->type) {
+            case 'rocketchat':
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Notifications/Model/Channel.php:71]: Now the channel type can be anything (user defined), and `HasAddress` cannot be changed accordingly.

We had a discussion that the behaviour can be removed, as it is only used to create icons.  They have recommended removing `has_rc` but keeping `has_email` for now as it is the most commonly used.

Since rc is provided here by default, I have added the icon.

File: application/forms/ChannelForm.php (hunk 5 above)
Comment (sukhwinder33445) [reply application/forms/ChannelForm.php:198]: The fieldset object must first be added as an element so that all children can use the decorator.

File: application/forms/ChannelForm.php [hunk 10]
```diff
@@ -229,35 +167,150 @@ protected function onSuccess()
...
+     *
+     * Fallback to locale `en_US` if the current locale isn't provided
+     *
+     * @param stdClass $localeMap
+     *
+     * @return ?string Only returns null if the fallback locale is also not specified
+     */
+    protected function fromCurrentLocale(stdClass $localeMap): ?string
+    {
+        $default = "en_US";
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:311]: please use `StaticTranslator::getDefaultLocale()`

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]
